src/utils/signals.py

```diff
@@ -162,50 +162,37 @@ def cfs_freq_analysis(self, metadata):
+        tau = round(metadata["tau_spacing"] * 1e-6 * fs)  # puts into units of samples
+        pulses_in_samples = [int(round(p * tau)) for p in pulses]
+        pulse_dur = round(0.0006 * fs)  # TODO: determine way to derive 0.0006
+        start_sample = int(round(pulses_in_samples[0] + pulse_dur / 2))
+        if len(pulses_in_samples) > 1:
+            end_sample = int(round(pulses_in_samples[1] - pulse_dur / 2))
+        else:
+            end_sample = self.antennas_iq_samples.shape[-1]
+
+        # If FFT resolution should be determined first, then N calculated based on the rate
+        # n = int(round(fs / output_freq_resolution))
+
+        # If an N should be chosen and FFT resolution calculated based on N
+        n = 512  # An ideal size of fft
+        output_freq_resolution = fs / n
 
-        masks = []
-        for i, slice_metadata in enumerate(metadata):
-            pulses = slice_metadata["pulses"]
-            tau = round(
-                slice_metadata["tau_spacing"] * 1e-6 * fs
-            )  # puts into units of samples
-            pulses_in_samples = [int(round(p * tau)) for p in pulses]
-            pulse_dur = round(0.0006 * fs)
-            mask = np.zeros(data.shape[-1], dtype=bool)
-            for pulse in pulses_in_samples:
-                start_mask = int(round(pulse - pulse_dur / 2))
-                end_mask = int(round(pulse + pulse_dur / 2))
-                if start_mask < 0:
-                    start_mask = 0
-                if end_mask >= len(mask):
-                    end_mask = len(mask) - 1
-                mask[start_mask:end_mask] = 1
-            masks.append(mask)
-
-        for i in range(len(masks)):
-            data[i, :, masks[i]] = 0.0 + 0.0j
-            # Masks samples recorded during transmission pulses
-
-        masked_fft = np.sum(np.abs(fft.fftshift(fft.fft(data), axes=-1)), axis=0)
-
-        freqs = fft.fftshift(fft.fftfreq(data.shape[-1], d=1 / fs))
+        num_intervals = int((end_sample - start_sample) / n)
+        end_sample = start_sample + num_intervals * n
+        data = self.antennas_iq_samples[:, :, start_sample:end_sample]
+        data_chunks = np.reshape(data, data.shape[:-1] + (num_intervals, n))
+
+        fft_data = fft.fftshift(fft.fft(data_chunks, axis=-1), axes=-1)
+        freqs = fft.fftshift(fft.fftfreq(n, d=1 / fs))
         df = freqs[1] - freqs[0]
-        output_freq_resolution = 1000  # Hz
-        sample_resolution = int(round(output_freq_resolution / df))
-
-        kernel = np.ones((sample_resolution,)) / sample_resolution
-
-        reduced_fft = np.array(
-            [
-                np.convolve(masked_fft[i, :], kernel, mode="same")
-                for i in range(masked_fft.shape[0])
-            ]
-        )[..., ::sample_resolution]
-        cfs_data = reduced_fft
-        cfs_freqs = freqs[::sample_resolution] - df / 2
+
+        cfs_data = np.sum(np.abs(np.mean(fft_data, axis=2)), axis=1)
+        cfs_freqs = freqs - df / 2
```

Comment:
Why is this done? I'm not sure this is correct

Reply:
Comment unclear.

As I understand it:
1. The input `antennas_iq_data` has a shape of  `(2, 20, 30507)`. Note: in this example we have 2 cfs slices.
2. We calculate the range of data to use from the end of the first pulse (`start_sample = 100`) to the start of the next pulse. In this case there is no next pulse so `end_sample=30507`.
3. Then we truncate by calculating `num_intervals` based on the size of the fft `n`, getting now that after truncating `end_sample=30308`.
4. So we end up with using a set of data with the shape `(2, 20, 30208)`
5. We then reshape it to get a set of ffts to sample the frequency spectrum multiple times, ending up with data chunks with a size of  `(2, 20, 59, 512)`
6. Then take an ffts across the last axis to generate `fft_data`
7. So when we go to return the frequencies and data measured we return `freqs` shifted to base-band
8. `cfs_data` is a determined then as:
   a. Take the average or mean of the `fft_data` across axis index 2 to average together the 59 samples of the frequency spectrum to get a new array shape of `(2, 20, 512)`.
  b. Next take the absolute value of that array and then sum across all antennas on axis 1 of the output of the previous step to get a final `cfs_data` shape of `(2, 512)`.

The only problem I could see was maybe we should be doing an average instead of a mean, so I have switched it to do an average across the 59 ffts. Let me know if I missed anything.

Reply:
Oh woops I'm sorry, I was talking about line 195: `cfs_freqs = freqs - df / 2`. I wasn't sure why the frequencies are being adjusted down slightly. Shouldn't the frequencies be exactly correct already?

Reply:
Ah I see what you mean. I figured it was shifting to centre around zero, but you're right since `fftfreqshift` already does that. The `df/2` must have slipped in at some point while we were iterating on this. On the bright side the effect it has is almost unnoticeable (subtracting ~300Hz off KHz scale values). But I'll remove it since it is incorrect.